toptal/picasso
# feat(RichTextEditor): lazy load quill

[FX-2759]

### Description

`QuillEditor` component is lazy loaded so that instance of quill does not load when using Picasso without tree shaking.
This way will the bundle size on other project will decrease and it should work out of the box for frameworks with SSR when importing Picasso components from the root like this:

```js
import { Container } from '@toptal/picasso'
```

### How to test

First of all you can [check temploy](https://picasso.toptal.net/FX-2759-quill-load/?path=/story/components-richtexteditor--richtexteditor) that RichTextEditor is working as expected. Then you can check whether quill is imported by following the next steps:

- Create a local project in Next.js using [Picasso guide](https://picasso.toptal.net/?path=/story/tutorials-how-to-use-server-side-rendering--how-to-use-server-side-rendering) - look for **Next.js** section.
- Don't stub quill (last sentence in step 5.)
- import (one) Picasso component without tree-shaking in `pages/index.tsx`: 
```js
import { Container } from '@toptal/picasso'
```
- run `yarn dev` and open `http://localhost:3000/`

You should see an error indicating that quill has been imported and parsed on the server

![image](https://user-images.githubusercontent.com/22159416/169026906-1ca43ef6-f50a-4b0c-9c56-926e63a4abd5.png)

- now update Picasso version in package.json to alpha: `25.1.1-alpha-FX-2759-quill-load.0`
- run `yarn dev` and open `http://localhost:3000/`

Now it should work without an error
 
### Development checks


**Breaking change**


<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and

1. You have no extra requests.
2. You have optional requests.
   1. Add `nit:` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because

1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:

1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>


[FX-2759]: https://toptal-core.atlassian.net/browse/FX-2759?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
